### PR TITLE
Include coarse woody debris (CWD) in heterotrophic respiration (HR)

### LINKED ISSUE #1361: HR fluxes don't include CWD
Since CLM4 coarse woody debris physically fragments into `LIT2` and `LIT3` pools with zero respiration fraction.
`rf_cwdl2_bgc = 0.`  

As a result this HR flux will not be calculated if `rf_cwdl2_bgc > 0.` 
https://github.com/ESCOMP/CTSM/blob/61770ad07d6eeecc0b6c5e9447c95f7c63e605d5/src/soilbiogeochem/SoilBiogeochemCarbonFluxType.F90#L242

Maybe not a big deal with our current parameterization, but it causes balance check errors in the PPE runs that @djk2120 is running.

@dlawrenncar and @djk2120 do you want this fixed for the PPE, or should we table it for now?  Regardless, this may be something @slevisconsulting addresses in his soil BGC refactoring we're doing for the MIMICS project.

### LINKED ISSUE #1417: NoAnthro test fails when checking for stream_meshfile_popdens file as it's none
### Brief summary of bug

NoAnthro use test case fails in the build-namelist unit tester

> 406/1740 < FAIL> <Test Id: 406> <Desc: -res 0.9x1.25 -bgc bgc -use_case 1850_noanthro_control -drydep -fire_emis -light_res 360x720 -phys clm5_0: lnd_in file exists>

### General bug information

**CTSM version you are using:** ctsm5.1.dev044
**Does this bug cause significantly incorrect results in the model's science?**No
**Configurations affected:** NoAnthro use case

### Important details of your setup / configuration so we can reproduce the bug

It does list the file as none in 'Buildconf/ctsm.input_data_list, but check_input_data safely seems to ignore it.

### Important output or errors that show the problem

The error it shows is:

> CLM adding use_case 1850_noanthro_control defaults for var 'stream_meshfile_popdens' with val 'none'
> ERROR : CLM build-namelist::CLMBuildNamelist::add_default() : file not found: stream_meshfile_popdens = /glade/p/cesmdata/cseg/inputdata/none

### PR DESCRIPTION
**Pull request by slevisconsulting**
### Description of changes

- Removing if-statements limiting history output of HR to non-CWD pools and decomposition.
- Adding cwdhr_col variable to track this heterotrophic respiration the same way that we track lithr_col.

### Specific notes

Contributors other than yourself, if any:
@wwieder 
@ekluzek 

CTSM Issues Fixed (include github issue #):
 Fixes #1361 
 Fixes #1417 

Are answers expected to change (and if so in what way)?
Yes.
- Currently inactive history variables that haven't included CWD to date, now will include CWD.
- Introducing history variables CWDC_HR (for C12, C13, C14). These and other CWD fields pertaining to HR will equal zero with default parameter values `rf_cwdl*_bgc = 0`. They will be greater than zero and other history fields will change when `rf_cwdl*_bgc > 0`.

Any User Interface Changes (namelist or namelist defaults changes)?
No

Testing performed, if any:
Running cheyenne test-suite. Not done but this far tests are passing.

### CODE REVIEW

File: src/soilbiogeochem/SoilBiogeochemCarbonFluxType.F90 [hunk 1]
```diff
@@ -349,7 +345,7 @@ subroutine InitHistory(this, bounds, carbon_type)
 
         this%decomp_cpools_leached_col(begc:endc,:) = spval
         this%decomp_cpools_transport_tendency_col(begc:endc,:,:) = spval
-        do k = 1, ndecomp_pools
+        do k = 1, ndecomp_pools  ! none from CWD
            if ( .not. decomp_cascade_con%is_cwd(k) ) then
```
Comment (slevisconsulting) [inline src/soilbiogeochem/SoilBiogeochemCarbonFluxType.F90:349]: This if-statement pertains to leaching and there's a corresponding if-statement in `SoilBiogeochemNitrogenFluxType.F90`.

I did not remove these. Should I?

File: src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90 [hunk 2]
```diff
@@ -9,8 +9,7 @@ module SoilBiogeochemDecompCascadeBGCMod
   use shr_kind_mod                       , only : r8 => shr_kind_r8
   use shr_const_mod                      , only : SHR_CONST_TKFRZ
   use shr_log_mod                        , only : errMsg => shr_log_errMsg
-  use clm_varpar                         , only : nlevsoi, nlevgrnd
-  use clm_varpar                         , only : nlevdecomp, ndecomp_cascade_transitions, ndecomp_pools, ndecomp_pools_max
+  use clm_varpar                         , only : nlevdecomp, ndecomp_pools_max
```
Comment (slevisconsulting) [inline src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90:12]: This is just for clean-up.

File: src/soilbiogeochem/SoilBiogeochemDecompMod.F90 [hunk 3]
```diff
@@ -127,7 +127,6 @@ subroutine SoilBiogeochemDecomp (bounds, num_soilc, filter_soilc,
          w_scalar                         =>    soilbiogeochem_carbonflux_inst%w_scalar_col                           , & ! Input:  [real(r8) (:,:)   ]  fraction by which decomposition is limited by moisture availability
          decomp_cascade_hr_vr             =>    soilbiogeochem_carbonflux_inst%decomp_cascade_hr_vr_col               , & ! Output: [real(r8) (:,:,:) ]  vertically-resolved het. resp. from decomposing C pools (gC/m3/s)
          decomp_cascade_ctransfer_vr      =>    soilbiogeochem_carbonflux_inst%decomp_cascade_ctransfer_vr_col        , & ! Output: [real(r8) (:,:,:) ]  vertically-resolved het. resp. from decomposing C pools (gC/m3/s)
-         decomp_k                         =>    soilbiogeochem_carbonflux_inst%decomp_k_col                           , & ! Output: [real(r8) (:,:,:) ]  rate constant for decomposition (1./sec)
```
Comment (slevisconsulting) [inline src/soilbiogeochem/SoilBiogeochemDecompMod.F90:130]: Just for clean-up.

File: src/soilbiogeochem/SoilBiogeochemPotentialMod.F90 [hunk 4]
```diff
@@ -109,29 +109,18 @@ subroutine SoilBiogeochemPotential (bounds, num_soilc, filter_soilc, &
...
-         decomp_cascade_ctransfer_vr      => soilbiogeochem_carbonflux_inst%decomp_cascade_ctransfer_vr_col        , & ! Output: [real(r8) (:,:,:) ]  vertically-resolved het. resp. from decomposing C pools (gC/m3/s)
          decomp_k                         => soilbiogeochem_carbonflux_inst%decomp_k_col                           , & ! Output: [real(r8) (:,:,:) ]  rate constant for decomposition (1./sec)      
-         phr_vr                           => soilbiogeochem_carbonflux_inst%phr_vr_col                             , & ! Output: [real(r8) (:,:)   ]  potential HR (gC/m3/s)                           
-         fphr                             => soilbiogeochem_carbonflux_inst%fphr_col                                 & ! Output: [real(r8) (:,:)   ]  fraction of potential SOM + LITTER heterotrophic
+         phr_vr                           => soilbiogeochem_carbonflux_inst%phr_vr_col                               & ! Output: [real(r8) (:,:)   ]  potential HR (gC/m3/s)
```
Comment (slevisconsulting) [inline src/soilbiogeochem/SoilBiogeochemPotentialMod.F90:123]: Just for clean-up.

File: src/soilbiogeochem/SoilBiogeochemCarbonFluxType.F90 (hunk 1 above)
Comment (slevisconsulting) [reply src/soilbiogeochem/SoilBiogeochemCarbonFluxType.F90:349]: ...and I would like confirmation that I removed the correct if-statements in `SoilBiogeochemCarbonFluxType.F90` and `SoilBiogeochemNitrogenFluxType.F90`.

**Comment by slevisconsulting** [comment]
Both test suites OK 
One caveat @ekluzek which may be a known problem:

The scripts `cs.status.fails` and `cs.status` give an error, so I used the scripts `*gnu`, `*int`, `*pgi`, `*nag` to see the results.

File: src/soilbiogeochem/SoilBiogeochemCarbonFluxType.F90 (hunk 1 above)
Comment (slevisconsulting) [reply src/soilbiogeochem/SoilBiogeochemCarbonFluxType.F90:349]: One more `if (.not. is_cwd)` that I did NOT remove:

https://github.com/slevisconsulting/ctsm/blob/219c6c19e603b29fc74a80e66ee54c1e57ca0a4a/src/soilbiogeochem/SoilBiogeochemLittVertTranspMod.F90#L283

This one pertains to the handling of the spin-up. Again let's discuss whether it should be removed.

**Comment by slevisconsulting** [comment]
@wwieder I realize that next week is the LMWG, so no pressure on the code review!

Also I was thinking of folding the solution to #1392 into this PR. I will wait to discuss with you first.

**Comment by slevisconsulting** [comment]
@ekluzek is suggesting running a test, baseline and new code, with the relevant history fields active.

**Comment by slevisconsulting** [comment]
I ran a single test with the relevant history fields active. The test is successful (PASS). Here are some details:

To generate a new baseline, I changed 8 lines with default `inactive` to `active` in master; the four in SoilBiogeochemCarbonFluxType.F90 and the four in SoilBiogeochemNitrogenFluxType.F90 corresponding to my mods in my branch. The baseline that I generated had 32 new history fields not found in the original baseline (otherwise bfb).

I changed the same 8 lines in my branch, which resulted in an additional 8 history fields not found in my new baseline (otherwise bfb). The 8 fields are cwd-related as I would have expected:

```
Could not find match for file1 variable CWD_HR_L2 in file2
Could not find match for file1 variable CWD_HR_L2_vr in file2
Could not find match for file1 variable CWD_HR_L3 in file2
Could not find match for file1 variable CWD_HR_L3_vr in file2
Could not find match for file1 variable SMINN_TO_LITR2N_CWD in file2
Could not find match for file1 variable SMINN_TO_LITR2N_CWD_vr in file2
Could not find match for file1 variable SMINN_TO_LITR3N_CWD in file2
Could not find match for file1 variable SMINN_TO_LITR3N_CWD_vr in file2
```

@ekluzek was wondering whether we should keep the new fields (40 in total) as active from now on. This is a question for @wwieder .

@wwieder I was also hoping you'd confirm whether my mods seem correct to you in this PR.
Plus we were going to discuss #1392 and whether to fold that into this PR.
These are just reminders. I am happy to wait for a convenient time for you to respond and/or meet w me.

File: src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90 [hunk 5]
```diff
@@ -334,27 +334,25 @@ subroutine InitHistory(this, bounds)
 
     do l = 1, ndecomp_cascade_transitions
        ! vertically integrated fluxes
-       !-- mineralization/immobilization fluxes (none from CWD)
-       if ( .not. decomp_cascade_con%is_cwd(decomp_cascade_con%cascade_donor_pool(l)) ) then
```
Comment (wwieder) [inline src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90:338]: I'm a little stuck wondering what we should do here... I wasn't imagining this flux would be associated with N mineralization or immobilization at this point.  Maybe it should remain with the comment that there is none from CWD

File: src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90 [hunk 6]
```diff
@@ -371,27 +369,25 @@ subroutine InitHistory(this, bounds)
 
        ! vertically resolved fluxes
        if ( nlevdecomp_full > 1 ) then
-          !-- mineralization/immobilization fluxes (none from CWD)
-          if ( .not. decomp_cascade_con%is_cwd(decomp_cascade_con%cascade_donor_pool(l)) ) then
```
Comment (wwieder) [inline src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90:375]: as above, I think we still want to exclude mineral N fluxes from CWD

File: src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90 [hunk 7]
```diff
@@ -443,32 +439,28 @@ subroutine InitHistory(this, bounds)
 
     if (.not. use_nitrif_denitrif) then
        do l = 1, ndecomp_cascade_transitions
-          !-- denitrification fluxes (none from CWD)
-          if ( .not. decomp_cascade_con%is_cwd(decomp_cascade_con%cascade_donor_pool(l)) ) then
```
Comment (wwieder) [inline src/soilbiogeochem/SoilBiogeochemNitrogenFluxType.F90:447]: Again, let's leave CWD out of the inorganic N calculations.

File: src/soilbiogeochem/SoilBiogeochemPotentialMod.F90 (hunk 4 above)
Comment (wwieder) [reply src/soilbiogeochem/SoilBiogeochemPotentialMod.F90:123]: is this code just not being used?

**Review by wwieder — COMMENTED** [review]
few comments related to N here, @slevisconsulting that we can discuss if it's helpful.

we need this on the PPE branch, which makes me wonder what changes need to be made to the parallel matrix code for respiration fluxes of CWD.  @ekluzek can you advise

**Comment by wwieder** [comment]
@slevisconsulting I don't think we need these extra history fields turned on by default. We should check that that the simulations are bfb with `rf_cwd` = 0 (not sure what the parameter is actually called), but then HR fluxes are greater when `rf_cwd` > 0.

**Comment by ekluzek** [comment]
@slevisconsulting yes, I agree with @wwieder they shouldn't be default on. But, I do wonder if they should be turned on for some of our testing. You can add them to hist_fincl1 to the user_nl_clm in one of the test mods. 

I'm thinking maybe it should go in the cropMonthOutput test mod as Crop can only be on if one of the BGC modes is on (otherwise test mods can be used for both SP or BGC).

**Review by wwieder — APPROVED** [review]
is this ready to go @slevisconsulting if so, can @ekluzek take it to the PPE branch & main?

**Comment by slevisconsulting** [comment]
> [...] check that that the simulations are bfb with `rf_cwd` = 0 (not sure what the parameter is actually called), but then HR fluxes are greater when `rf_cwd` > 0.

With my current mods (as they appear in the PR), the test is bfb with `rf_cwdl2_bgc=0` and `rf_cwdl3_bgc=0`.

However, when I set `rf_cwdl2_bgc=0.5` and `rf_cwdl3_bgc=0.5`, I get a carbon balance error, for example:

```
column cbalance error    =   3.063005290509873E-006           8
 Latdeg,Londeg=   30.0000000000000        15.0000000000000
 begcb                    =    435.841365763495
 endcb                    =    435.840938319034
 delta store              =  -4.274444609109196E-004
 --- Inputs ---
 gpp                      =   0.000000000000000E+000
 --- Outputs ---
 er                       =   4.243814556363771E-004
 col_fire_closs           =   0.000000000000000E+000
 col_hrv_xsmrpool_to_atm  =   0.000000000000000E+000
 col_xsmrpool_to_atm      =   0.000000000000000E+000
 wood_harvestc            =   0.000000000000000E+000
 grainc_to_cropprodc      =   0.000000000000000E+000
 -1*som_c_leached         =   1.596739547638851E-014
```

I have not looked into it, yet. Let me know if you have ideas. And @wwieder you mentioned something about the parallel matrix code. Any chance this error relates to your comment?

**Comment by ekluzek** [comment]
@slevisconsulting and @wwieder don't worry about the parallel matrix code for now. That's the work on the PPE branch that we might need to provide for when it's moved over to there. Let's get this version working and onto master. When I move it to the PPE branch, I might need some help applying the same changes for the matrix, but possibly not.

Also @slevisconsulting we should add a test mod for this test and add it to the testlist. We'll want to make sure this feature is working for future changes (and for example on the PPE branch).  The testmod can point to the params file with the changes you had to make to test it out.

**Comment by wwieder** [comment]
sorry, bouncing between meetings w/o much time, but did we ever include CWD_HR in the balance check equation?  Presumably this needs to be included in the heterotrophic respiration (HR) flux?

**Comment by slevisconsulting** [comment]
> [...] did we ever include CWD_HR in the balance check equation? Presumably this needs to be included in the heterotrophic respiration (HR) flux?

This is what I'm finding. In fact there is no `cwdhr_col` variable the way there is `lithr_col` and `somhr_col`.

**Comment by slevisconsulting** [comment]
I will add `cwdhr_col` using `lithr_col` as template.

**Comment by wwieder** [comment]
Thanks for checking @slevisconsulting.  Funny that it was in the title for the PR and the issue, but not in the code :)

**Comment by slevisconsulting** [comment]
Yes, I misunderstood #1361 as being about the fluxes missing from history. Sorry about that...

File: src/biogeochem/CNVegCarbonFluxType.F90 [hunk 8]
```diff
@@ -282,7 +282,6 @@ module CNVegCarbonFluxType
      real(r8), pointer :: xsmrpool_recover_patch                    (:)     ! (gC/m2/s) C flux assigned to recovery of negative cpool
      real(r8), pointer :: xsmrpool_c13ratio_patch                   (:)     ! C13/C(12+13) ratio for xsmrpool (proportion)
 
-     real(r8), pointer :: cwdc_hr_col                               (:)     ! (gC/m2/s) col-level coarse woody debris C heterotrophic respiration
```
Comment (slevisconsulting) [inline src/biogeochem/CNVegCarbonFluxType.F90:285]: `cwdc_hr_col` unnecessary.
I introduced `cwdhr_col` instead, following `lithr_col` as template.

**Comment by slevisconsulting** [comment]
> [...] check that that the simulations are bfb with `rf_cwd` = 0 (not sure what the parameter is actually called), but then HR fluxes are greater when `rf_cwd` > 0.

- My latest commit fixes the C balance error.
- I have confirmed in a single test that simulations remain bfb when `rf_cwdl*_bgc=0` and that the new history variable `CWDC_HR` = 0 everywhere.
- I have confirmed using the same test that answers change when `rf_cwdl*_bgc=0.5` and that `CWDC_HR` > 0.

To finish this work, I need to address @ekluzek 's recommendation:

> add a test mod for this test and add it to the testlist. We'll want to make sure this feature is working for future changes (and for example on the PPE branch). The testmod can point to the params file with the changes you had to make to test it out.

**Comment by slevisconsulting** [comment]
> @slevisconsulting yes, I agree with @wwieder they shouldn't be default on. But, I do wonder if they should be turned on for some of our testing. You can add them to hist_fincl1 to the user_nl_clm in one of the test mods.
> 
> I'm thinking maybe it should go in the cropMonthOutput test mod as Crop can only be on if one of the BGC modes is on (otherwise test mods can be used for both SP or BGC).

Before

**Comment by wwieder** [comment]
one more naive question here @slevisconsulting are there any additional changes that need to be included for summing `CWDC_HR` fluxes related to isotopes, or is this taken care of? 

If so, let's see what @ekluzek says about testing.

**Comment by slevisconsulting** [comment]
I discussed the new test with @ekluzek, and we converged on this new test:
`ERS_Ld3.f10_f10_mg37.I2000Clm51Bgc.cheyenne_intel.clm-cwdhr`
where the new testmod `cwdhr` will add to the default testmod the line `hist_fincl1 = 'CWDC_HR','CWD_HR_L2','CWD_HR_L2_vr','CWD_HR_L3','CWD_HR_L3_vr','SMINN_TO_LITR2N_CWD','SMINN_TO_LITR2N_CWD_vr','SMINN_TO_LITR3N_CWD','SMINN_TO_LITR3N_CWD_vr'`

@wwieder I haven't done anything with isotopes in this PR, so I will take a look.

**Comment by slevisconsulting** [comment]
...oh, unless you mean C12 vs. C13 vs. C14. I did take care of those.
@wwieder pls let me know if I'm confused.

**Comment by ekluzek** [comment]
@slevisconsulting if this effects C13 and C14, your test case likely should turn them on as well. There are some Carbon isotope test mods like "ciso", you might want to have your testmod inherit from it, and then keep ciso in the name. So something like ciso_cwd_hr...

**Comment by wwieder** [comment]
yes, let's make sure the carbon isotopes C13 & C14 tests pass as expected (and are tracking the CWD_HR).

**Comment by slevisconsulting** [comment]
> yes, let's make sure the carbon isotopes C13 & C14 tests pass as expected (and are tracking the CWD_HR).

I ran the new test `ERS_Ld3.f10_f10_mg37.I2000Clm51Bgc.cheyenne_intel.clm-ciso_cwd_hr` in two ways:
1. pointing to the default params file where rf_cwd* = 0
2. pointing to my alternate params file where rf_cwd* = 0.5
In the first case the CWD fields associated with HR (incl. C13 & C14) were zero everywhere.
In the second case the same fields contained values greater than zero.

I have run the cheyenne and izumi test-suites and they are OK.
I have run ./rimport to get the new params file in the repo.

@wwieder @ekluzek 
after a final review of the PR, this work is ready for merge.
Pls let me know of any concerns.
Also let me know when I'm next in line, so that I may rerun the test-suites.

**Comment by wwieder** [comment]
no additional suggestions here. looks good to me!

**Review by ekluzek — APPROVED** [review]

**Pull request merged**